src/Symfony/Component/Translation/Bridge/Phrase/PhraseProvider.php

```diff
@@ -0,0 +1,305 @@
+                ]);
+
+                if (200 !== ($statusCode = $response->getStatusCode()) && 304 !== $statusCode) {
+                    $this->logger->error(sprintf('Unable to get translations for locale "%s" from phrase: "%s".', $locale, $response->getContent(false)));
+
+                    $this->throwProviderException($statusCode, $response, 'Unable to get translations from phrase.');
+                }
+
+                $content = 304 === $statusCode && null !== ($cache = $this->getCache($item)) ? $cache->getContent() : $response->getContent();
+                $translatorBag->addCatalogue($this->loader->load($content, $locale, $domain));
+
+                // using weak etags, responses for requests with fallback locale enabled can not be reliably cached...
+                if (false === $this->readConfig->isFallbackLocaleEnabled()) {
+                    $headers = $response->getHeaders(false);
+                    $item->set(new PhraseCachedResponse($headers['etag'][0], $headers['last-modified'][0], $content));
+                    $this->cache->save($item);
+                }
+            }
+        }
+
+        $event = new PhraseReadEvent($translatorBag);
+        $this->dispatcher->dispatch($event);
+
+        return $event->getBag();
+    }
+
+    public function write(TranslatorBagInterface $translatorBag): void
+    {
+        \assert($translatorBag instanceof TranslatorBag);
+
+        $event = new PhraseWriteEvent($translatorBag);
+        $this->dispatcher->dispatch($event);
+
+        $catalogues = $event->getBag()->getCatalogues();
+
+        foreach ($catalogues as $catalogue) {
+            foreach ($catalogue->getDomains() as $domain) {
+                if (0 === \count($catalogue->all($domain))) {
+                    continue;
+                }
+
+                $phraseLocale = $this->getLocale($catalogue->getLocale());
+
+                $content = $this->xliffFileDumper->formatCatalogue($catalogue, $domain, ['default_locale' => $this->defaultLocale]);
+                $filename = sprintf('%d-%s-%s.xlf', date('YmdHis'), $domain, $catalogue->getLocale());
+
+                $fields = array_merge($this->writeConfig->withTag($domain)->withLocale($phraseLocale)->getOptions(), ['file' => new DataPart($content, $filename, 'application/xml')]);
+
+                $formData = new FormDataPart($fields);
+
+                $response = $this->httpClient->request('POST', 'uploads', [
+                    'body' => $formData->bodyToIterable(),
+                    'headers' => $formData->getPreparedHeaders()->toArray(),
+                ]);
+
+                if (201 !== $statusCode = $response->getStatusCode()) {
+                    $this->logger->error(sprintf('Unable to upload translations for domain "%s" to phrase: "%s".', $domain, $response->getContent(false)));
+
+                    $this->throwProviderException($statusCode, $response, 'Unable to upload translations to phrase.');
+                }
```

Comment:
i would move this to a second foreach outside of the foreaches for catalogues and domains to enable concurrent http requests:
https://symfony.com/doc/current/http_client.html#concurrent-requests

btw, we could consider to insolate the entire HTTP and caching concern into a `PhraseClient` or something to slim down the provider class and it's dependencies

Reply:
agreed with the concurrent requests. 
a dedicated client is an interesting idea, have to look into that...

Reply:
unfortunately phrase only allows 4 concurrent requests
https://developers.phrase.com/api/#overview--rate-limiting

as chunking the response array doesn't wait, we could work with promises, but then the question would be: is the additional dependency worth the benefit (and also do you want a mandatory non-symfony dependency)..

i suppose that that in most cases the uploaded diff is fairly small so the response time should be as well..
any thoughts on that?

Reply:
wasn't aware of that limitation. let's keep it this way!